### PR TITLE
[WIP]    Linux/Android native deploy

### PR DESCRIPTION
**Pull request by PariksheetPinjari909**
For https://github.com/dmlc/tvm/issues/973

### CODE REVIEW

**Comment by tqchen** [comment]
@PariksheetPinjari909 this example uses the C API and requires JNI. Can you build a version that uses the java API instead of JNI? The Fix for thread runtime is useful and we can bring that in first in a separate PR. Thanks!

Specifically, we can:
- Build a library.so with export_library
- Build TVM java runtime without RPC support
- Use TVM java runtime to run the prediction in java

**Comment by tqchen** [comment]
The native JNI deployment can also be helpful, but it would be great to show a java API example, since that is much easier for the users. Also if we are only running prediction example, we don't need to have extra library to save arrays, simply print out the class id and possible class would do for a demo app

**Comment by tqchen** [comment]
cc @yzhliu

**Comment by PariksheetPinjari909** [comment]
Ref: "Talk about how to use the java API": Yes, build with JAVA makes it much easier than JNI for android app developers. Working in it 

Ref: " we don't need to have extra library to save arrays, simply print out the class id and possible class would do for a demo app"

- NPY format for in & out was with an intention of developers first try with Python for initial experiments on X86 then bring it to target.
- Input varies like direct JPEG stream, decoded Tensor (resized, converted to YUV domain ...etc.)

So to keep the deploy tool generic I preferred npy format. NPY format dump is a single addition to initial python script to dump input and to feed model output back for processing after prediction.

**Comment by tqchen** [comment]
Usually, app developer wants something that is more stand alone. Like take an image and how could I feed it to the app. We can build an example app that takes an image(from camera or file) do a preprocessing and feed it to java runtime. See https://github.com/dmlc/nnvm/blob/master/tutorials/web/resnet.html#L80 for a javascript version of preprocessing

**Comment by PariksheetPinjari909** [comment]
Android Standalone demo application tutorial updated.

**Comment by tqchen** [comment]
@yzhliu can you review this?

**Comment by janboeye** [comment]
@PariksheetPinjari909 
How to native deploy tvm.build for just one layer? Is it same?

Thanks

File: tutorials/deployment/native/deploy_native.cc [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    cnpy::NpyArray in_arr = cnpy::npy_load(model_input + ".npy");
+    float* loaded_data = in_arr.data<float>();
+
+    DLTensor* x;
+    int in_ndim = in_arr.shape.size();
+    int64_t *in_shape = (int64_t *) &in_arr.shape[0];
+    uint64_t num_samples=1;
+    TVMArrayAlloc(in_shape, in_ndim, dtype_code, dtype_bits, dtype_lanes, kDLCPU, device_id, &x);
+
+    // Initlalize the Input
```
Comment (yzhliu) [inline tutorials/deployment/native/deploy_native.cc]: typo 'Initlalize'

File: docs/how_to/deploy_android.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+
+TVM Runtime for android Target
+-------------------------------
+
+This is a cross compilation process of libtvm_runtime.so for android with OpenCL support.
+
+
+Prerequisites:
+Android stand alone tool chain : ANDROID_NDK_PATH/ndk/android-toolchain-arm64/bin/aarch64-linux-android-
```
[the remark this answers is not in the document]
Comment (srkreddy1238) [reply docs/how_to/deploy_android.md]: Path is complete, it's the tool chain prefix I showed there.
Explaining detailed steps to produce tool chain need not be here. Any I added the android URL to refer it.

[the remark this answers is not in the document]
Comment (PariksheetPinjari909) [reply docs/how_to/deploy_android.md]: path is complete only, it's tool chain prefix reference.

About standalone tool chain building: I think it's out of scope for current context. Any way I added android URL reference to build it.

File: docs/how_to/deploy_android.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+│       ├── cl_gl.h
+│       ├── cl.h
+│       ├── cl_platform.h
+│       ├── cl_va_api_media_sharing_intel.h
+│       └── opencl.h
+└── lib
+    └── libOpenCL.so
+```
+
+- Enable OPENCL in make/config.mk
```
[the remark this answers is not in the document]
Comment (PariksheetPinjari909) [reply docs/how_to/deploy_android.md]: Done. Added  required explanation for this.

File: docs/how_to/deploy_android.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+Above example will directly run the compiled model on RPC target. Below modification at [rum_mobile.py](https://github.com/tkat0/chainer-nnvm-example/blob/5b97fd4d41aa4dde4b0aceb0be311054fb5de451/run_mobile.py#L64) will save the compilation output which is required on android target.
+
+```
+lib.export_library("YOLOv2_tiny-aarch64.so", ndk.create_shared)
+
+with open("YOLOv2_tiny-aarch64.json", "w") as fo:
+        fo.write(graph.json())
+with open("YOLOv2_tiny-aarch64.params", "wb") as fo:
+        fo.write(nnvm.compiler.save_param_dict(params))
+```
```
[the remark this answers is not in the document]
Comment (PariksheetPinjari909) [reply docs/how_to/deploy_android.md]: Done.

File: docs/how_to/deploy_android.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+- Enable OPENCL in make/config.mk
+
+- Below command can build libtvm_runtime.so for android.
+
+```
+OPENCL_PATH=<path to OOPENCL path> CXX=<ANDROID_NDK_PATH>/ndk/android-toolchain-arm64/bin/aarch64-linux-android-g++ make lib/libtvm_runtime.so
+```
+
+Result of this setp is the libtvm_runtime.so which is a dependency for Android native while building application.
```
[the remark this answers is not in the document]
Comment (PariksheetPinjari909) [reply docs/how_to/deploy_android.md]: Done.

File: tutorials/deployment/native/README.txt [hunk 6]
```diff
@@ -0,0 +1,62 @@
+/*!
```
[the remark this answers is not in the document]
Comment (PariksheetPinjari909) [reply tutorials/deployment/native/README.txt]: Done.

**Comment by PariksheetPinjari909** [comment]
@janboeye yes the procedure is same for one layer too.

**Comment by tqchen** [comment]
Thanks for the changes, let us move ```tutorials/deployment/android -> apps/android_deploy```

File: apps/android_deploy/app/src/main/jni/tvm_runtime.h [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+#include "../src/runtime/cpu_device_api.cc"
+#include "../src/runtime/workspace_pool.cc"
+#include "../src/runtime/module_util.cc"
+#include "../src/runtime/system_lib_module.cc"
+#include "../src/runtime/module.cc"
+#include "../src/runtime/registry.cc"
+#include "../src/runtime/file_util.cc"
+#include "../src/runtime/dso_module.cc"
+#include "../src/runtime/thread_pool.cc"
+
```
Comment (tqchen) [inline apps/android_deploy/app/src/main/jni/tvm_runtime.h]: add #include "../src/runtime/thread_backend.cc" Please rebase against the master and test the workflow

File: docs/how_to/deploy_android.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
+How to deploy and use compiled model on Android
+===============================================
+
+This tutorial explain below aspects (Unlike the android_rpc approach we already have)
+
+    1: Build a model for android target.
+    2: TVM runtime building for android target.
+    3: A sample native application to test the model on ADB before integrating into android application (apk).
```
Comment (tqchen) [inline docs/how_to/deploy_android.md]: I think we should first refer to the apps/deploy_android on how to use java API, the native API comes second

**Comment by tqchen** [comment]
@PariksheetPinjari909 Let us remove the native deployment tutorial for now as it is in largely duplicates with the apps/how_to_deploy. Focusing on the java api example and make it clear and good is useful for most users

We can still keep a section on what is needed to be done if user prefers JNI and refers most of example code to apps/how_to_deploy.

**Comment by tqchen** [comment]
Thanks for the patient during the reviewing process and keep improving the PR, this is now merged

**Pull request merged**